Clearable time selector

schelv opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue or discussion
  in the additional information section.
-->
The time selector can now be cleared:
![image](https://github.com/home-assistant/frontend/assets/13403863/228f7b8d-699f-4d48-ad50-e5ffc98cc053)

The same selector after pressing the clear button:
![image](https://github.com/home-assistant/frontend/assets/13403863/89a135d4-235f-4845-8d5f-df12116d2db8)

## Type of change
<!--
  What type of change does your PR introduce to the Home Assistant frontend?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR.
-->

```yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue or discussion:
- Link to documentation pull request:

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

<!--
  Thank you for contributing <3
-->

[docs-repository]: https://github.com/home-assistant/home-assistant.io


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced a clear button for time input components, allowing users to easily clear the selected time. The clear button is conditionally displayed based on specific properties (`clearable`, `required`, and `disabled`).

- **Enhancements**
  - Updated time input components to support the `clearable` property for improved user control over input fields, including visual and functional updates to accommodate the new feature.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

schelv committed:
initial attempt at clearable time selector

schelv committed:
only show clear button if there is a value

KTibow commented on src/components/ha-time-input.ts:
```diff
@@ -60,7 +63,9 @@ export class HaTimeInput extends LitElement {
     let value;
 
-    if (
+    if (typeof eventValue === "undefined") {
```
why do this instead of just prepending `eventValue &&` to the existing checks

schelv replied:
I did not think of that option yet.
Pressing the "clear" button is an explicit action, and therefore it makes sense (to me) to explicitly set `value` to `undefined` as well. I think it makes the code a bit easier to understand.

But i'm not really satisfied with the "undefined" here, because it does not show the intention here (why the value is undefined).
Do you know if there is better approach? one that more clearly conveys the message "this value is being cleared"

KTibow replied:
no, it's not about clearing the value, the value is already cleared
it's about a type safety check
as such something like this makes more sense
```js
if (
  eventValue &&
  (!isNaN(eventValue.hours) || !isNaN(eventValue.minutes) || !isNaN(eventValue.seconds))
) {
  // ...
}
```

schelv replied:
I've replaced `undefined` with `null`, since it "represents the intentional absence of any object value", which is what I am aiming for.
I've also added a comment for clarification.
Thank you for the feedback!😀

bramkragten commented on src/components/ha-base-time-input.ts:
```diff
@@ -249,13 +253,28 @@ export class HaBaseTimeInput extends LitElement {
+    const value = undefined;
+
+    fireEvent(this, "value-changed", {
+      value,
```
This is not allowed according to the current typing (that should have been checked here)

schelv replied:
It looks like it is not necessarily a `TimeChangedEvent`
![image](https://github.com/home-assistant/frontend/assets/13403863/0f9492b9-c537-4187-9c9a-00724b2955e0)

schelv replied:
I changed the typing of the `_timeChanged` method👍

schelv committed:
use null instead of undefined.

bramkragten commented on src/components/ha-base-time-input.ts:
```diff
@@ -249,13 +253,26 @@ export class HaBaseTimeInput extends LitElement {
 
+  private _clearValue(): void {
+    fireEvent(this, "value-changed", {
+      value: null,
```
Why `null` and `undefined`?
```suggestion
      value: undefined,
```

schelv replied:
I added a comment to the place where this is "caught". I can add it here as well.

bramkragten replied:
I don't think the consumer should deal with this, the input value is just cleared, so undefined.

schelv replied:
Ideally there would be some kind of value-cleared event that can be used.

Having a value-changed event without a value can be confusing.
So I googled a bit and using [null](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/null) seems like a reasonable option here.

schelv replied:
I've removed the `null` value

schelv committed:
leave event value undefined

Quentame reviewed (approved):
LGTM

karwosts commented:
Once we get this would be good to add for `duration` selector as well, can fix #19360

bdraco commented:
I think this will solve the issue where its not possible to clear a timeout without dropping back to YAML once its been set which seems like the root cause of https://github.com/home-assistant/core/issues/109586

It also resolves quandary in https://github.com/home-assistant/core/pull/115830 as it means it means we align the frontend to the backend 

- no timeout set (cleared), no timeout happens
- 0 timeout set, times out right away

bdraco committed:
Merge branch 'dev' into clearable-time-selector

schelv commented:
@bramkragten I see you added the "Needs UX" label. Does that mean that some action needs to be taken for this merge request?

matthiasdebaat commented:
@schelv thanks for your PR! We discussed this because it adds a nice extra feature to the input, but having the `X` outside the input is not up to standards. With a couple of tweaks, I think we can have the `X` inside the input.

Based on the [M3 textfields](https://m3.material.io/components/text-fields/guidelines), [M3 time pickers](https://m3.material.io/components/time-pickers) and [NN/g guidelines](https://www.nngroup.com/articles/date-input/), I would suggest the following design. More details in the [Figma File](https://www.figma.com/design/ciR6Xr2ryjaOakWEA38vgc/Material-3-Design-Kit-(Community)?node-id=57797-2818&t=LWYqONepVKkVI62q-1). I made it with the M3 Figma Kit, so the colors are wrong.

![CleanShot 2024-07-03 at 11 54 28@2x](https://github.com/home-assistant/frontend/assets/20203284/7ebfdb1b-e94c-44e8-add2-2e36baa83de1)

### Changes

- Move the AM/PM dropdown to a selector next to the input.
- Show hour, minute and second labels when there is a value
- Make the input label [adjacent](https://m3.material.io/components/text-fields/guidelines#a2c6b301-66b1-412f-991a-bf9b7ad0a012).
- When empty, have a placeholder. Default can be `No time`,  but I'm open for suggestions.

schelv committed:
Merge branch 'dev' into clearable-time-selector

schelv committed:
move clear button into input

schelv commented:
@matthiasdebaat
I've moved the `X` to be inside of the input, added labels, and separated the dropdown.  I'm not the right person to implement the rest of the design. I have no idea what I'm doing.😅

![image](https://github.com/user-attachments/assets/2676b162-88f4-48af-97cc-58b766431b2c)

matthiasdebaat commented:
It's already a great improvement, thanks! Can you extend this line?
![CleanShot 2024-07-15 at 11 09 19@2x](https://github.com/user-attachments/assets/2d0881b4-0e3a-407c-9ac4-12e4ee9a6fb6)

schelv commented:
> It's already a great improvement, thanks! Can you extend this line? ![CleanShot 2024-07-15 at 11 09 19@2x](https://private-user-images.githubusercontent.com/20203284/348652147-2d0881b4-0e3a-407c-9ac4-12e4ee9a6fb6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjEwMzUyOTYsIm5iZiI6MTcyMTAzNDk5NiwicGF0aCI6Ii8yMDIwMzI4NC8zNDg2NTIxNDctMmQwODgxYjQtMGUzYS00MDdjLTlhYzQtMTJlNGVlOWE2ZmI2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzE1VDA5MTYzNlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk4M2Q3MjUwNGE2NGVkOWZhMzlkMjdhNDg3Mjg4NWY3ZDk1MTNhMWQ4NDE2Y2I5YmQ3NDJlOGFkYmZkOGI4ZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IkR3OeVQ4Ft5PkTYKgdGJ4twdmd0qHCV5F0NOmyee0E)

I tried to do that but could not find which css property is responsible for that line.

schelv committed:
add line

schelv commented:
@matthiasdebaat 
![image](https://github.com/user-attachments/assets/6941316a-178e-402e-9f7f-c8f30e15079f)
🎉

matthiasdebaat commented:
YOU ROCK! Thanks!

silamon commented:
@matthiasdebaat , the current ux is good to go?

matthiasdebaat commented:
> @matthiasdebaat , the current ux is good to go?

One question, before it's good to go. Does this field have a [adjacent](https://m3.material.io/components/text-fields/guidelines#a2c6b301-66b1-412f-991a-bf9b7ad0a012) label? Or do you need to add one above by yourself?

bramkragten commented on src/components/ha-time-input.ts:
```diff
@@ -48,22 +50,26 @@ export class HaTimeInput extends LitElement {
     const eventValue = ev.detail.value;
 
     const useAMPM = useAmPm(this.locale);
     let value;
```
```suggestion
    let value: string | undefined;
```

schelv committed:
Update src/components/ha-time-input.ts

Co-authored-by: Bram Kragten <mail@bramkragten.nl>

silamon commented on src/components/ha-base-time-input.ts:
```diff
@@ -77,27 +79,27 @@ export class HaBaseTimeInput extends LitElement {
    * Label for the day input
    */
-  @property() dayLabel = "";
+  @property() dayLabel = "Day";
```
These should be translatable.

schelv replied:
Is this the way to do it?
`this.hass!.localize("ui.components.<something?>")`

silamon replied:
Yeah, the translations should be added to https://github.com/home-assistant/frontend/blob/dev/src/translations/en.json
I would suggest ui.components.selectors.time.<day/...>. After that you can use them in the code that way. 

Please note that this.hass may be uninitialized at the time you're assigning dayLabel and so on. You may need to remove the default value so if not defined it ends up as being undefined and then check within the render if the dayLabel is defined and if not, use this.hass.localize("label") to fill in the default. 😃

schelv replied:
I've taken the easy way out and made the labels empty again.
I declare it out of scope 😄

passing around hass objects that may or may not exist is a little bit too much for this merge request.

silamon replied:
Ok, it shouldn't be much passing around, I'll take a look if I can add it.

schelv replied:
Do you want to add it to this pull request or to a fresh one?

silamon replied:
I just had a look and it's a component that doesn't have access to hass. So, let's declare it out of scope. 
They should be passed in with the properties by the components that do have access to hass.

silamon reviewed (commented):
These assignments should be translatable.

silamon commented on src/components/ha-base-time-input.ts:
```diff
@@ -302,18 +318,22 @@ export class HaBaseTimeInput extends LitElement {
+    }
     :host {
-      display: block;
+      display: flex;
```
Before vs now: 
![image](https://github.com/user-attachments/assets/ba726c32-6141-4134-9e36-a96a1704c408)
 vs
![image](https://github.com/user-attachments/assets/1baec510-5d26-479b-bb5d-dd5e2d7f8209)

Change this back to display: block so the label and the input fields appear again in 2 rows. 
then wrap the input fields in a new div and make that display: flex so the flex css isn't applied to the label.

schelv replied:
I did not really consider the label until now.
So far I've been using blueprints to test the selector, and I can't figure out how to set the label from the blueprint itself.

In order to test the label positioning I hardcoded the label value and this is how it looks:
"normal" window size:
![image](https://github.com/user-attachments/assets/312969a9-8678-466a-8ea9-0c54ce35c101)
and smaller window size:
![image](https://github.com/user-attachments/assets/07191d91-879e-49a8-a19e-810385cb58e9)

I'm guessing the label is used in a non-blueprint context.
But why doesn't the blueprint use it as well?

If we want to changed it however, it is best to do it in another pull request. =)

silamon replied:
This component is used in a lot of different places next to the blueprint editor, you will also encounter it in scripts, automations, config flows and parts of the UI.

schelv committed:
Merge branch 'dev' into clearable-time-selector

schelv committed:
improve label appearance

schelv committed:
remove the label values
